lib/common/busypoll.c

```diff
@@ -0,0 +1,444 @@
+    netdev_napi_set_req_free(set_req);
+    ynl_sock_destroy(ys);
+}
+
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof((arr)[0]))
+#endif
+
+void attach_cbpf(int fd, uint16_t cpus, const char *iface)
+{
+    struct sockaddr_storage ss;
+    socklen_t sslen = sizeof(ss);
+    uint16_t n = cpus;
+    int ret = 0;
+
+    ret = getsockname(fd, (struct sockaddr *)&ss, &sslen);
+    if (ret != 0) {
+        fprintf(stderr, "getsockname failed on listener: %s\n", strerror(errno));
+    }
+    if (ss.ss_family == AF_UNIX) {
+        fprintf(stderr, "dont add reuseport filter to AF_UNIX socket listener\n");
+        return;
+    } else if (ss.ss_family == AF_INET) {
+        fprintf(stderr, "will add filter to AF_INET socket\n");
+    } else if (ss.ss_family == AF_INET6) {
+        fprintf(stderr, "will add filter to AF_INET6 socket\n");
+    } else {
+        fprintf(stderr, "unknown family: %lu, adding\n", (unsigned long)ss.ss_family);
+    }
+
+    struct sock_filter code[] = {
+        /* A = skb->queue_mapping */
+        {BPF_LD | BPF_W | BPF_ABS, 0, 0, SKF_AD_OFF + SKF_AD_QUEUE},
+        /* A = A % n */
+        {BPF_ALU | BPF_MOD, 0, 0, n},
+        /* return A */
+        {BPF_RET | BPF_A, 0, 0, 0},
+    };
+
+    struct sock_fprog p = {
+        .len = ARRAY_SIZE(code),
+        .filter = code,
+    };
+
+    if (setsockopt(fd, SOL_SOCKET, SO_ATTACH_REUSEPORT_CBPF, &p, sizeof(p))) {
+        char buf[128];
+        h2o_fatal("failed to set SO_ATTACH_REUSEPORT_CBPF: %d:%s", errno, h2o_strerror_r(errno, buf, sizeof(buf)));
+    } else {
+        fprintf(stderr, "bpf prog attached to fd:%d, queues: %u\n", fd, cpus);
+    }
+
+    return;
+}
+
+void bind_interface(int fd, const char *iface)
+{
+    /* setting SO_BINDTODEVICE causes listen sockets with the same
+     * attributes (address, port, nic index) to be part of the same
+     * reuseport group.
+     *
+     * this is critical because if the nic index is not set with
+     * BINDTODEVICE, the reuseport group will receive incoming connections
+     * from any NIC and the cBPF program which does a modulo on the
+     * queue_mapping will fail (since multiple NICs each can have the same
+     * queue mapping)
+     */
+    if (iface && setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, iface, strlen(iface)) == -1) {
+        h2o_perror("failed to SO_BINDTODEVICE");
+        abort();
+    }
+
+    if (iface) {
+        fprintf(stderr, "bound %d to iface: %s\n", fd, iface);
+    }
+}
+
+static unsigned int get_napi_id(int fd)
+{
+#if !defined(SIOCGIFNAME_BY_NAPI_ID)
+#define SIOCGIFNAME_BY_NAPI_ID (0x894D)
+#endif
+    unsigned int napi_id = 0;
+    socklen_t napi_id_len = sizeof(napi_id);
+
+    if (getsockopt(fd, SOL_SOCKET, SO_INCOMING_NAPI_ID, &napi_id, &napi_id_len) != 0) {
+        h2o_perror("so_napi_incoming_id failed");
+    }
+
+    return napi_id;
+}
+
+static int get_nic_name_by_napi(int fd, unsigned int napi_id, struct ifreq *ifr)
```

Comment:
this function (and its callers) should be re-written -- this ioctl is just a hack, not a thing that exists in the real public linux kernel.

instead, when h2o starts it should compute the mapping of interfaces and NAPI IDs, and the implementation of this function would take a NAPI ID and return the interface name from the map h2o has computed

Reply:
thank you, i rewrote `get_nic_name_by_napi`. im at loss at how to reimplement `get_napi_id`, but im thinking we don't need to if we save the napi_id into the loop at assign time. WDYT?